### ISSUE TITLE
Page title doesn't take OS' filepath length limit into account

### ISSUE THREAD

**Issue by GiantCrocodile** [getkirby/kirby#1821]
You can create a page title with a very long name and this will result in an `mkdir(): No such file or directory` error. Obviously the maximum filepath length limit isn't taken into account while checking the title of a page.

Not sure how relevant this issue is - maybe it's relevant for API requests or virtual pages. I just spotted it by chance. Watch out for the length limits of Windows 10 and Linux, as they are different as you most likely know.

**Comment by afbora** [comment]
I have an idea 💡 

We can use `PHP_MAXPATHLEN` constant.
But this setting is ignored on some platforms (Apache on Windows Machine).
So we need also manual control with `PHP_OS` constant.

Rewritten `Dir::make()` method

https://github.com/getkirby/kirby/blob/master/src/Toolkit/Dir.php#L210-L233

```php
public static function make(string $dir, bool $recursive = true): bool
{
    if (empty($dir) === true) {
        return false;
    }

    if (is_dir($dir) === true) {
        return true;
    }

    $parent = dirname($dir);

    if ($recursive === true) {
        if (is_dir($parent) === false) {
            static::make($parent, true);
        }
    }

    if (is_writable($parent) === false) {
        throw new Exception(sprintf('The directory "%s" cannot be created', $dir));
    }

   // Check file length
    try {
        return mkdir($dir);
    } catch (Exception $e) {
        $dirLength = strlen($dir);

        switch (PHP_OS) {
            // Max file length is 255 on Windows
            // We should check out FAT, NTFS file systems
            case "WINNT":
                if ($dirLength > 255) {
                    throw new Exception(sprintf('The directory name exceeded limit: %d and cannot be created ', 255));
                }
                break;
            default:
                if ($dirLength > PHP_MAXPATHLEN) {
                    throw new Exception(sprintf('The directory name exceeded limit: %d and cannot be created ', PHP_MAXPATHLEN));
                }
                break;
        }
    }
}
```

**Comment by GiantCrocodile** [comment]
@afbora Would this be something for Kirby 3.4.1? I'm asking because in the future this issue could be more relevant when more people start to create pages by API calls from third party solutions/services and thus don't think about checking length of page title to truncate it if necessary.

**Comment by afbora** [comment]
@GiantCrocodile I would love to, but it is a situation that differs according to the OS. 
This can may be resolved by checking title limit like with `page.create:before` 🤔

**Comment by GiantCrocodile** [comment]
Actually it wouldn't be even worst if we just set one Kirby limit, regardless of what the OS does support. An unified length would be helpful in general I think. And honestly, I think nobody cares if let's say it's 200 or 250 chars; just a proper error handling would be great so the developer knows there's a limit and issue to be fixed ongoing.

**Comment by texnixe** [comment]
You can check the title length inside a model using the create method. 
```
    public static function create(array $props): Page
    {
        
        if (Str::Length($props['content']['title']) > 50) {
            throw new Exception('The title is longer than 50 characters');
        } 
        
        return parent::create($props);
    }
```
But in general it would be useful if we could limit title length via a blueprint setting although I wonder if there are really people who would choose endlessly long titles.

**Comment by afbora** [comment]
The @getkirby/kirby-staff should make a general limitation decision, but you can use the hook I said as a workaround: `page.create:before`

**Comment by lukasbestle** [comment]
The title itself can be as long as the user wants, the only limitation is the length of the slug. So I think the cleanest solution would be to limit the slug length to something manageable. If no custom slug is given, the auto-generated slug could be cut automatically. If a custom slug is too long, the creation would fail with an Exception.

**Comment by afbora** [comment]
Implemented @lukasbestle idea on #2709 and waiting your reviews.

**Comment by bastianallgeier** [comment]
✅
